node/heartbeat/monitor.go

```diff
@@ -92,46 +165,62 @@ func (m *Monitor) SetAppStatusHandler(ash core.AppStatusHandler) error {
 // ProcessReceivedMessage satisfies the p2p.MessageProcessor interface so it can be called
 // by the p2p subsystem each time a new heartbeat message arrives
 func (m *Monitor) ProcessReceivedMessage(message p2p.MessageP2P) error {
-	if message == nil || message.IsInterfaceNil() {
-		return ErrNilMessage
-	}
-	if message.Data() == nil {
-		return ErrNilDataToProcess
+	hbRecv, err := m.messageHandler.CreateHeartbeatFromP2pMessage(message)
+	if err != nil {
+		return err
 	}
 
-	hbRecv := &Heartbeat{}
+	//message is validated, process should be done async, method can return nil
+	go m.addHeartbeatMessageToMap(hbRecv)
 
-	err := m.marshalizer.Unmarshal(hbRecv, message.Data())
-	if err != nil {
-		return err
+	return nil
+}
+
+func (m *Monitor) addHeartbeatMessageToMap(hb *Heartbeat) {
+	m.mutHeartbeatMessages.Lock()
+	defer m.mutHeartbeatMessages.Unlock()
+
+	pubKeyStr := string(hb.Pubkey)
+	hbmi, ok := m.heartbeatMessages[pubKeyStr]
+	if hbmi == nil || !ok {
+		var err error
+		hbmi, err = newHeartbeatMessageInfo(m.maxDurationPeerUnresponsive, false, m.genesisTime, m.timer.Now)
+		if err != nil {
+			log.Error(err.Error())
+			return
+		}
+		m.heartbeatMessages[pubKeyStr] = hbmi
 	}
 
-	err = m.verifySignature(hbRecv)
+	computedShardID := m.computeShardID(pubKeyStr)
+	hbmi.HeartbeatReceived(computedShardID, hb.ShardID, hb.VersionNumber, hb.NodeDisplayName)
+	hbDTO := m.convertToExportedStruct(hbmi)
+	err := m.storer.SavePubkeyData(hb.Pubkey, &hbDTO)
 	if err != nil {
-		return err
+		log.Warn(fmt.Sprintf("cannot save heartbeat to db: %s", err.Error()))
 	}
+	m.addPeerToFullPeersSlice(hb.Pubkey)
+	//m.updateAllHeartbeatMessages()
```

Comment:
commented code. Might remove it as the update is being done by the GetHeartbeats func

Reply:
yes! fixed